### PR TITLE
Refactor to allow configurable backends (configmap, eks configmap, crd)

### PR DESCRIPTION
**Pull request by wongma7**
This enables the authenticator server to support multiple backends (sources of IAM->user+group mappings) at the same time with a well-defined order of precedence according to the backend-mode argument.

Fixes https://github.com/kubernetes-sigs/aws-iam-authenticator/issues/247
also fixes https://github.com/kubernetes-sigs/aws-iam-authenticator/issues/261

TODO: more tests -- will do in a follow-up
~~TODO: update README~~ -- done
~~TODO: fix server_test.go~~ -- done
TODO: sts regional endpoint -- needs some discussion, will do in a follow-up

### CODE REVIEW

**Comment by nckturner** [comment]
@wongma7 awesome, yeah I think the regional endpoint is a separate concern.

File: cmd/aws-iam-authenticator/server.go [hunk 1]
```diff
@@ -139,12 +123,16 @@ func init() {
...
 
-	serverCmd.Flags().String(
-		"address",
+	serverCmd.Flags().String("address",
 		"127.0.0.1",
 		"IP Address to bind the server to listen to. (should be a 127.0.0.1 or 0.0.0.0)")
 	viper.BindPFlag("server.address", serverCmd.Flags().Lookup("address"))
 
+	serverCmd.Flags().StringSlice("backend-mode",
+		[]string{"ConfigMap", "File"},
```
Comment (christopherhein) [inline cmd/aws-iam-authenticator/server.go]: is this missing `crd`? Also should these be Capitalized? Maybe stored and a `const` and referenced?

Comment (wongma7) [reply cmd/aws-iam-authenticator/server.go]: yes, added crd & replaced with const

File: cmd/aws-iam-authenticator/server.go [hunk 2]
```diff
@@ -139,12 +123,16 @@ func init() {
...
-	serverCmd.Flags().String(
-		"address",
+	serverCmd.Flags().String("address",
 		"127.0.0.1",
 		"IP Address to bind the server to listen to. (should be a 127.0.0.1 or 0.0.0.0)")
 	viper.BindPFlag("server.address", serverCmd.Flags().Lookup("address"))
 
+	serverCmd.Flags().StringSlice("backend-mode",
+		[]string{mapper.ModeCRD, mapper.ModeConfigMap, mapper.ModeFile},
+		fmt.Sprintf("Ordered list of backends to get mappings from. Comma-delimited list of: %s", strings.Join(mapper.BackendModeChoices, ",")))
```
Comment (micahhausler) [inline cmd/aws-iam-authenticator/server.go]: Can you specify what "ordered" means in this context? First one that matches wins? Last one wins?

File: pkg/mapper/configmap/configmap.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+	"k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/aws-iam-authenticator/pkg/config"
+)
+
+type MapStore struct {
+	mutex sync.RWMutex
+	users map[string]config.UserMapping
+	roles map[string]config.RoleMapping
+	// TODO: Use kubernetes set.
```
Comment (micahhausler) [inline pkg/mapper/configmap/configmap.go]: Is this something we want to do now?

Comment (wongma7) [reply pkg/mapper/configmap/configmap.go]: not much advantage to doing so as we don't do any set operations (TODO was left by someone else). Will leave TODO intact though.

File: cmd/aws-iam-authenticator/server.go (hunk 2 above)
Comment (wongma7) [reply cmd/aws-iam-authenticator/server.go]: first one wins, added clarification.

**Comment by wongma7** [comment]
I think tests can be a follow-up. But I need to update the documentation (and fix the travis build). Will ping when that is ready.

File: pkg/mapper/configmap/configmap.go (hunk 3 above)
Comment (nckturner) [reply pkg/mapper/configmap/configmap.go]: If we agree that there is no advantage, let's remove the TODO to reduce confusion.

File: pkg/mapper/configmap/configmap.go [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+	if err != nil {
+		return nil, err
+	}
+	clientset, err := kubernetes.NewForConfig(clientconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	ms := MapStore{}
+	// TODO: Should we use a namespace?  Make it configurable?
```
Comment (nckturner) [inline pkg/mapper/configmap/configmap.go]: Make a constant for now and remove the TODO.

File: pkg/mapper/configmap/configmap.go [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+				case watch.Error:
+					logrus.WithFields(logrus.Fields{"error": r}).Error("recieved a watch error")
+				case watch.Deleted:
+					logrus.Info("Resetting configmap on delete")
+					userMappings := make([]config.UserMapping, 0)
+					roleMappings := make([]config.RoleMapping, 0)
+					awsAccounts := make([]string, 0)
+					ms.saveMap(userMappings, roleMappings, awsAccounts)
+				case watch.Added, watch.Modified:
+					// Type assertion is not working
```
Comment (nckturner) [inline pkg/mapper/configmap/configmap.go]: Remove this comment.

File: pkg/mapper/configmap/configmap.go [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+	errors []error
+}
+
+func (err ErrParsingMap) Error() string {
+	return fmt.Sprintf("error parsing config map: %v", err.errors)
+}
+
+// Acquire lock before calling
+func (ms *MapStore) parseMap(m map[string]string) ([]config.UserMapping, []config.RoleMapping, []string, error) {
+	// TODO: Look at errors.Wrap().
```
Comment (nckturner) [inline pkg/mapper/configmap/configmap.go]: Remove.

File: pkg/mapper/configmap/configmap.go [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+	awsAccounts := make([]string, 0)
+	if accountsData, ok := m["mapAccounts"]; ok {
+		err := yaml.Unmarshal([]byte(accountsData), &awsAccounts)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	// TODO: Check for empty user and role mappings.
```
Comment (nckturner) [inline pkg/mapper/configmap/configmap.go]: And do what?

File: README.md [hunk 8]
```diff
@@ -84,7 +84,7 @@ systemctl restart kubelet.service
...
+In the `master` version of the AWS IAM Authenticator you can configure your users using one of three methods. The `mapUsers` and `mapRoles` as seen in the [Full Configuration Format](#full-configuration-format), using the new (alpha) [Kubernetes Custom Resource Definitions](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/), or using an [EKS aws-auth ConfigMap](https://docs.aws.amazon.com/eks/latest/userguide/add-user-role.html). The `--backend-mode` flag determines which of these methods is enabled and their order of precedence (first one wins). The CRD and ConfigMap methods allow the authenticator server to always stay upto date with the latest allowed user. See [Issues #79](https://github.com/kubernetes-sigs/aws-iam-authenticator/issues/79) for more details.
```
[the remark this answers is not in the document]
Comment (wongma7) [reply README.md]: backup?

File: pkg/mapper/configmap/configmap.go [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+				case watch.Deleted:
+					logrus.Info("Resetting configmap on delete")
+					userMappings := make([]config.UserMapping, 0)
+					roleMappings := make([]config.RoleMapping, 0)
+					awsAccounts := make([]string, 0)
+					ms.saveMap(userMappings, roleMappings, awsAccounts)
+				case watch.Added, watch.Modified:
+					switch cm := r.Object.(type) {
+					case *core_v1.ConfigMap:
+						// TODO: Only watch on configmap/awsauth
```
[the remark this answers is not in the document]
Comment (wongma7) [reply pkg/mapper/configmap/configmap.go]: seems it's already done by lines above
`FieldSelector: fields.OneTermEqualSelector("metadata.name", "aws-auth").String(),`

File: pkg/mapper/configmap/mapper.go [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+	return &ConfigMapMapper{ms}, nil
+}
+
+func (m *ConfigMapMapper) Name() string {
+	return mapper.ModeConfigMap
+}
+
+func (m *ConfigMapMapper) Start(_ <-chan struct{}) error {
+	// TODO respect stopCh
+	m.startLoadConfigMap()
```
[the remark this answers is not in the document]
Comment (wongma7) [reply pkg/mapper/configmap/mapper.go]: yes. i made a point not to touch any configmap code but this is a no-brainer.

**Comment by wongma7** [comment]
RE: tests there are a couple tests that inadvertently use multiple mappers but no explicit tests yet. https://github.com/kubernetes-sigs/aws-iam-authenticator/pull/269/files#diff-8151a8a5ca3b510f06b426f0cfb48708R646 I would like to refactor server_test.go in a followup as it will be quite verbose to have a configmap+file+crd version of every testcase. Then add some explicit tests.

**Pull request merged**